Build failure: arrow-cpp (staging branch)

### Steps To Reproduce

At **staging** (not `master`):
1. nix build .#arrow-cpp

### Build log

```
--   Outputting build configuration summary to /build/apache-arrow-16.0.0/cpp/build/cmake_summary.json
-- Configuring done (13.8s)
CMake Error at src/arrow/CMakeLists.txt:592 (target_link_libraries):
  Target "arrow_util" links to:

    LZ4::lz4

  but the target was not found.  Possible reasons include:

    * There is a typo in the target name.
    * A find_package call is missing for an IMPORTED target.
    * An ALIAS target is missing.
```
Full log: https://termbin.com/i0tg

### Additional context

(I've opened this to log or look after a solution for this.)

* For debug purposes, reverting #318512 fixes `arrow-cpp` build.

### Notify maintainers

CC @tobim @veprbl @cpcloud

--- a/pkgs/tools/compression/lz4/0001-Create-a-unified-lz4-target.patch
+++ b/pkgs/tools/compression/lz4/0001-Create-a-unified-lz4-target.patch
@@ -1,0 +1,34 @@
+From 14fca2e93bd783c0ad2362af6d69801c2073765d Mon Sep 17 00:00:00 2001
+From: Tobias Mayer <tobim@fastmail.fm>
+Date: Sun, 23 Jun 2024 20:16:00 +0200
+Subject: [PATCH] Create a unified lz4 target
+
+Includes changes from
+* https://github.com/lz4/lz4/pull/1372
+* https://github.com/lz4/lz4/pull/1413
+---
+ build/cmake/CMakeLists.txt | 8 ++++++++
+ 1 file changed, 8 insertions(+)
+
+diff --git a/build/cmake/CMakeLists.txt b/build/cmake/CMakeLists.txt
+index eb7007b..07ddd2b 100644
+--- a/build/cmake/CMakeLists.txt
++++ b/build/cmake/CMakeLists.txt
+@@ -130,6 +130,14 @@ if(BUILD_STATIC_LIBS)
+     POSITION_INDEPENDENT_CODE ${LZ4_POSITION_INDEPENDENT_LIB})
+   list(APPEND LZ4_LIBRARIES_BUILT lz4_static)
+ endif()
++# Add unified target.
++add_library(lz4 INTERFACE)
++list(APPEND LZ4_LIBRARIES_BUILT lz4)
++if(BUILD_SHARED_LIBS)
++  target_link_libraries(lz4 INTERFACE lz4_shared)
++else()
++  target_link_libraries(lz4 INTERFACE lz4_static)
++endif()
+ 
+ if(BUILD_STATIC_LIBS)
+   set(LZ4_LINK_LIBRARY lz4_static)
+-- 
+2.45.1
+

--- a/pkgs/tools/compression/lz4/default.nix
+++ b/pkgs/tools/compression/lz4/default.nix
@@ -1,4 +1,4 @@
-{ lib, stdenv, fetchFromGitHub, fetchpatch, cmake
+{ lib, stdenv, fetchFromGitHub, cmake
 , valgrind, testers
 }:
 
@@ -24,11 +24,7 @@ stdenv.mkDerivation (finalAttrs: {
   outputs = [ "dev" "lib" "man" "out" ];
 
   patches = [
-    (fetchpatch { # https://github.com/lz4/lz4/pull/1162
-      name = "build-shared-no.patch";
-      url = "https://github.com/lz4/lz4/commit/851ef4b23c7cbf4ceb2ba1099666a8b5ec4fa195.patch";
-      hash = "sha256-P+/uz3m7EAmHgXF/1Vncc0uKKxNVq6HNIsElx0rGxpw=";
-    })
+    ./0001-Create-a-unified-lz4-target.patch
   ];
 
   cmakeDir = "../build/cmake";

--- a/pkgs/tools/compression/lz4/default.nix
+++ b/pkgs/tools/compression/lz4/default.nix
@@ -57,5 +57,6 @@ stdenv.mkDerivation (finalAttrs: {
     license = with licenses; [ bsd2 gpl2Plus ];
     platforms = platforms.all;
     mainProgram = "lz4";
+    maintainers = [ maintainers.tobim ];
   };
 })

